### PR TITLE
Handle different environments

### LINKED ISSUE #47: Differentiate between used environments such as development env and production env.
### Story
We -developers- face errors, and when that occurs we need full information about the error and its stack trace. 
But our users should NOT have any kind or any information about any error if occurred. 
So we need to differentiate between the environment we use as developers, and between the environment that will be available for our users.

### Prerequisites
- [ ] Understand difference between dev env and prod env.
- [ ] Understand why do we need to have several env and what each env is supposed to present/have on certain actions, e.g. when an error occurs, what should logically each env need to present/show.

### Tasks
- [ ] Create development environment
- [ ] Create production environment

### Acceptance criteria
- [ ] Should be able to change between the needed environments. 
- [ ] Each one should have its correct logic presenting what's needed from it.

### PR DESCRIPTION
**Pull request by amroakmal**
Separated and limited the information from the generated error that should be returned/responded by, depending on the working environment.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
     "build": "nest build",
     "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
     "start": "nest start",
-    "start:dev": "nest start --watch",
+    "start:dev": "NODE_ENV=development nest start --watch",
```
Comment (Sahl-A) [inline package.json:13]: This is redundant. As you use development environment variable in the code base in app.module.ts line 15

File: src/app.module.ts [hunk 2]
```diff
@@ -7,12 +7,12 @@ import { PostsModule } from './posts/posts.module';
...
 import config from './config/database';
 
-const evnVariable = process.env.NODE_ENV || 'development';
+const evnVariable = process.env.NODE_ENV;
 @Module({
   imports: [
     ConfigModule.forRoot({
       // load different .env files based on runtime environment variable
-      envFilePath: [`.${evnVariable}.env`],
+      envFilePath: !evnVariable ? '.development.env' : `.${evnVariable}.env.`,
```
Comment (Sahl-A) [inline src/app.module.ts]: using development in line 15 instead of line 10 didn't change the functionality. Please clarify the reason of using it.

**Review by Sahl-A — CHANGES_REQUESTED** [review]
In addition to the comments in the code, I have the following considerations:
- When testing with pickify-posts-pr-56.pickify.net, The detailed error still returned. 
- The commits does not represent the changes done. Please consider this in future PRs

File: src/app.module.ts (hunk 2 above)
Comment (amroakmal) [reply src/app.module.ts]: My thought was, even If there has been a new environment that we will use later, or even if no env. variable for the environment is given, this should be redirected to the development assignment, i.e. making the development environment the default thing.

File: package.json (hunk 1 above)
Comment (amroakmal) [reply package.json:13]: Yeah, but the .env file is not shared with all of us.
So, if someone will follow-up and continue using this, the default environment that will be used is the development one.

File: src/logging/http-exceptions-logger.filter.ts [hunk 3]
```diff
@@ -12,6 +12,38 @@ import { formatISO } from 'date-fns';
 export class AllExceptionsFilterLogger implements ExceptionFilter {
   constructor(private readonly logger: Logger) {}
 
+  generateErrorBasedOnCurrentEnvironment(
```
Comment (Sahl-A) [inline src/logging/http-exceptions-logger.filter.ts]: - Add **generateErrorBasedOnCurrentEnvironment** function in a separate file. Also make it pure function, by passing NODE_ENV as a variable. Making it pure function will make test much easier.
- Add tests for the function

**Review by Sahl-A — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (amroakmal) [reply package.json:13]: @Sahl-A 
I figured out that we will need to explicitly add ```NODE_ENV=development``` in the command of  ```"start:dev"```
The following use-case occurred with me:
Sometimes I would explicitly need to know the name of the environment I'm currently at, for example in the ```generateErrorBasedOnCurrentEnvironment``` function.
Because depending on the current env I'm working at, I'll do specific actions., because If there's no env name sent for me, I will not be able to distinguish between whether nothing was sent for me or I was given a wrong env name?

P.S: I know we currently need only production and development environments, but I think in the future(or to be more generic) we may need others such as test environment.

Comment (Sahl-A) [reply package.json:13]: lgtm

**Review by Sahl-A — CHANGES_REQUESTED** [review]
- the changes look good to me.
- testing the function is missing
- I cannot test the deployed version: https://pickify-posts-pr-56.pickify.net/api/
  - I get 522 error.

File: src/shared/utils/LoggingUtils/generateErrorBasedOnEnvironment.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      message: message,
+      stack,
+      timestamp: getNow(),
+      path: url,
+    };
+  } else {
+    if (status >= 500) {
+      message = 'Internal Server Error';
+    }
+
```
Comment (saifaldin) [inline src/shared/utils/LoggingUtils/generateErrorBasedOnEnvironment.ts]: Why is that necessary ?
Also, if it is necessary, it doesn't change the status code.

Comment (amroakmal) [reply src/shared/utils/LoggingUtils/generateErrorBasedOnEnvironment.ts]: Because I shouldn't reveal any information about the error when its code is >= 500.
It changes the message that will be send to the user, and its effect will be shown when its merged.
Or do you mean something else?

Comment (saifaldin) [reply src/shared/utils/LoggingUtils/generateErrorBasedOnEnvironment.ts]: Ok but then you will need to change the status code too

Comment (amroakmal) [reply src/shared/utils/LoggingUtils/generateErrorBasedOnEnvironment.ts]: IMO, I don't think sending the real status code to the user back in the production environment.
The error -if occurred-, we should return the least possible information to the user, yet descriptive that an error occurred without showing any details.
What do you think?

**Review by Sahl-A — APPROVED** [review]
LGTM
